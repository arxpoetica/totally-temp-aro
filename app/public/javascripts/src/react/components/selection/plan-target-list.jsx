import React, { Component } from 'react'
import { PropTypes } from 'prop-types'
import reduxStore from '../../../redux-store'
import wrapComponentWithProvider from '../../common/provider-wrapped-component'
import SelectionActions from '../selection/selection-actions'
import SelectionModes from '../selection/selection-modes'

export class PlanTargetList extends Component {
  render () {
    var geometries = []
    var geometryKey = null
    var descriptionKey = null
    if (this.props.activeSelectionModeId === SelectionModes.SELECTED_LOCATIONS) {
      geometryKey = 'locations'
      descriptionKey = 'address'
      geometries = this.props.planTargets[geometryKey]
    } else if (this.props.activeSelectionModeId === SelectionModes.SELECTED_AREAS) {
      geometryKey = 'serviceAreas'
      descriptionKey = 'code'
      geometries = this.props.planTargets[geometryKey]
    } else if (this.props.activeSelectionModeId === SelectionModes.SELECTED_ANALYSIS_AREAS) {
      geometryKey = 'analysisAreas'
      descriptionKey = 'code'
      geometries = this.props.planTargets[geometryKey]
    }

    if (geometries.size === 0) {
      return <div>(no items selected)</div>
    } else {
      return <div>
        <div style={{ backgroundColor: '#e0e0e0', paddingLeft: '10px' }}>
          <span>{geometries.size} items selected</span>
          <button className='btn btn-outline-danger btn-sm float-right'
            style={{ marginTop: '3px' }}
            onClick={() => this.props.removePlanTargets(this.props.planId, { [geometryKey]: this.props.planTargets[geometryKey] })}
          >
            <i className='far fa-trash-alt' />
          </button>
        </div>
        <ul style={{ listStyleType: 'none', paddingLeft: '0px', maxHeight: '200px', overflowY: 'auto', marginBottom: '0px' }}>
          {
            [...geometries].map(item => {
              const description = this.props.planTargetDescriptions[geometryKey][item]
                ? this.props.planTargetDescriptions[geometryKey][item][descriptionKey]
                : 'loading...'
              return <li key={item}>
                <button className='btn btn-thin btn-outline-danger'
                  style={{ border: 'none' }}
                  onClick={() => this.props.removePlanTargets(this.props.planId, { [geometryKey]: new Set([item]) })}
                >
                  <i className='far fa-trash-alt' />
                </button>
                <span style={{ verticalAlign: 'middle' }}>{description}</span>
              </li>
            })
          }
        </ul>
      </div>
    }
  }
}

PlanTargetList.propTypes = {
  planId: PropTypes.number,
  activeSelectionModeId: PropTypes.string,
  planTargets: PropTypes.object,
  planTargetDescriptions: PropTypes.object
}

const mapStateToProps = (state) => ({
  planId: state.plan.activePlan && state.plan.activePlan.id,
  activeSelectionModeId: state.selection.activeSelectionMode.id,
  planTargets: state.selection.planTargets,
  planTargetDescriptions: state.selection.planTargetDescriptions
})

const mapDispatchToProps = (dispatch, ownProps) => ({
  removePlanTargets: (planId, planTargets) => dispatch(SelectionActions.removePlanTargets(planId, planTargets))
})

const PlanTargetListComponent = wrapComponentWithProvider(reduxStore, PlanTargetList, mapStateToProps, mapDispatchToProps)
export default PlanTargetListComponent