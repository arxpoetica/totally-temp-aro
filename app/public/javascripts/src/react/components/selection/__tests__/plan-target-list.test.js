/* global test expect jest */
import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import { PlanTargetList } from '../plan-target-list'
import SelectionModes from '../../selection/selection-modes'

// -----------------------------------------------------------------------------
test('Render when all plan targets are empty', () => {
  const component = renderer.create(
    <PlanTargetList planId={1}
      activeSelectionModeId={SelectionModes.SELECTED_LOCATIONS}
      planTargets={{ locations: new Set(), serviceAreas: new Set(), analysisAreas: new Set() }}
      planTargetDescriptions={{}}
    />
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

// -----------------------------------------------------------------------------
test('Render when location targets are being shown', () => {
  const planTargets = {
    locations: new Set([100, 200, 999]),
    serviceAreas: new Set(),
    analysisAreas: new Set()
  }

  const planTargetDescriptions = {
    locations: {
      100: { address: 'Test address 1' },
      999: { address: 'Test address 2' } // Note that we are intentionally keeping id 200 blank
    },
    serviceAreas: {},
    analysisAreas: {}
  }

  const component = renderer.create(
    <PlanTargetList planId={1}
      activeSelectionModeId={SelectionModes.SELECTED_LOCATIONS}
      planTargets={planTargets}
      planTargetDescriptions={planTargetDescriptions}
    />
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

// -----------------------------------------------------------------------------
test('Render when service area targets are being shown', () => {
  const planTargets = {
    locations: new Set(),
    serviceAreas: new Set([100, 200, 999]),
    analysisAreas: new Set()
  }

  const planTargetDescriptions = {
    locations: {},
    serviceAreas: {
      100: { code: 'STTLWA09' },
      200: { code: 'STTLWAXX' } // Note that we are intentionally keeping id 999 blank
    },
    analysisAreas: {}
  }

  const component = renderer.create(
    <PlanTargetList planId={1}
      activeSelectionModeId={SelectionModes.SELECTED_AREAS}
      planTargets={planTargets}
      planTargetDescriptions={planTargetDescriptions}
    />
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

// -----------------------------------------------------------------------------
test('Render when analysis area targets are being shown', () => {
  const planTargets = {
    locations: new Set(),
    serviceAreas: new Set(),
    analysisAreas: new Set([100, 200, 999])
  }

  const planTargetDescriptions = {
    locations: {},
    serviceAreas: {},
    analysisAreas: {
      100: { code: 'Seattle' },
      200: { code: 'More seattle' } // Note that we are intentionally keeping id 999 blank
    }
  }

  const component = renderer.create(
    <PlanTargetList planId={1}
      activeSelectionModeId={SelectionModes.SELECTED_ANALYSIS_AREAS}
      planTargets={planTargets}
      planTargetDescriptions={planTargetDescriptions}
    />
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

// -----------------------------------------------------------------------------
// Helper function to test removal of plan targets
const testRemovingPlanTarget = (targetKey, selectionMode) => {
  const planTargets = {
    locations: new Set(),
    serviceAreas: new Set(),
    analysisAreas: new Set()
  }
  planTargets[targetKey] = new Set([100, 200, 999])

  const planTargetDescriptions = {
    locations: {},
    serviceAreas: {},
    analysisAreas: {}
  }

  // Then try to delete one plan target
  const mockRemovePlanTargets = jest.fn()
  var component = shallow(
    <PlanTargetList planId={1}
      activeSelectionModeId={selectionMode}
      planTargets={planTargets}
      planTargetDescriptions={planTargetDescriptions}
      removePlanTargets={mockRemovePlanTargets}
    />
  )
  component.find('div > ul > li > button').at(0).simulate('click')
  expect(mockRemovePlanTargets.mock.calls[0][0]).toEqual(1)
  var setOfTargets = mockRemovePlanTargets.mock.calls[0][1][targetKey]
  expect(setOfTargets.has(100)).toBeTruthy()
  expect(setOfTargets.size).toEqual(1)

  // Now delete all the plan targets
  component = shallow(
    <PlanTargetList planId={1}
      activeSelectionModeId={selectionMode}
      planTargets={planTargets}
      planTargetDescriptions={planTargetDescriptions}
      removePlanTargets={mockRemovePlanTargets}
    />
  )
  component.find('div > div > button').simulate('click')
  expect(mockRemovePlanTargets.mock.calls[1][0]).toEqual(1)
  setOfTargets = mockRemovePlanTargets.mock.calls[1][1][targetKey]
  expect(setOfTargets.has(100)).toBeTruthy()
  expect(setOfTargets.has(200)).toBeTruthy()
  expect(setOfTargets.has(999)).toBeTruthy()
  expect(setOfTargets.size).toEqual(3)
}

// -----------------------------------------------------------------------------
test('Ability to remove location targets - single and multiple', () => {
  testRemovingPlanTarget('locations', SelectionModes.SELECTED_LOCATIONS)
})

// -----------------------------------------------------------------------------
test('Ability to remove service area targets - single and multiple', () => {
  testRemovingPlanTarget('serviceAreas', SelectionModes.SELECTED_AREAS)
})

// -----------------------------------------------------------------------------
test('Ability to remove analysis area targets - single and multiple', () => {
  testRemovingPlanTarget('analysisAreas', SelectionModes.SELECTED_ANALYSIS_AREAS)
})
