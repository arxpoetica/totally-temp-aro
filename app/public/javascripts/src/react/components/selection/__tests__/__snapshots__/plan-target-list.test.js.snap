// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render when all plan targets are empty 1`] = `
<div>
  (no items selected)
</div>
`;

exports[`Render when analysis area targets are being shown 1`] = `
<div>
  <div
    style={
      Object {
        "backgroundColor": "#e0e0e0",
        "paddingLeft": "10px",
      }
    }
  >
    <span>
      3
       items selected
    </span>
    <button
      className="btn btn-outline-danger btn-sm float-right"
      onClick={[Function]}
      style={
        Object {
          "marginTop": "3px",
        }
      }
    >
      <i
        className="far fa-trash-alt"
      />
    </button>
  </div>
  <ul
    style={
      Object {
        "listStyleType": "none",
        "marginBottom": "0px",
        "maxHeight": "200px",
        "overflowY": "auto",
        "paddingLeft": "0px",
      }
    }
  >
    <li>
      <button
        className="btn btn-thin btn-outline-danger"
        onClick={[Function]}
        style={
          Object {
            "border": "none",
          }
        }
      >
        <i
          className="far fa-trash-alt"
        />
      </button>
      <span
        style={
          Object {
            "verticalAlign": "middle",
          }
        }
      >
        Seattle
      </span>
    </li>
    <li>
      <button
        className="btn btn-thin btn-outline-danger"
        onClick={[Function]}
        style={
          Object {
            "border": "none",
          }
        }
      >
        <i
          className="far fa-trash-alt"
        />
      </button>
      <span
        style={
          Object {
            "verticalAlign": "middle",
          }
        }
      >
        More seattle
      </span>
    </li>
    <li>
      <button
        className="btn btn-thin btn-outline-danger"
        onClick={[Function]}
        style={
          Object {
            "border": "none",
          }
        }
      >
        <i
          className="far fa-trash-alt"
        />
      </button>
      <span
        style={
          Object {
            "verticalAlign": "middle",
          }
        }
      >
        loading...
      </span>
    </li>
  </ul>
</div>
`;

exports[`Render when location targets are being shown 1`] = `
<div>
  <div
    style={
      Object {
        "backgroundColor": "#e0e0e0",
        "paddingLeft": "10px",
      }
    }
  >
    <span>
      3
       items selected
    </span>
    <button
      className="btn btn-outline-danger btn-sm float-right"
      onClick={[Function]}
      style={
        Object {
          "marginTop": "3px",
        }
      }
    >
      <i
        className="far fa-trash-alt"
      />
    </button>
  </div>
  <ul
    style={
      Object {
        "listStyleType": "none",
        "marginBottom": "0px",
        "maxHeight": "200px",
        "overflowY": "auto",
        "paddingLeft": "0px",
      }
    }
  >
    <li>
      <button
        className="btn btn-thin btn-outline-danger"
        onClick={[Function]}
        style={
          Object {
            "border": "none",
          }
        }
      >
        <i
          className="far fa-trash-alt"
        />
      </button>
      <span
        style={
          Object {
            "verticalAlign": "middle",
          }
        }
      >
        Test address 1
      </span>
    </li>
    <li>
      <button
        className="btn btn-thin btn-outline-danger"
        onClick={[Function]}
        style={
          Object {
            "border": "none",
          }
        }
      >
        <i
          className="far fa-trash-alt"
        />
      </button>
      <span
        style={
          Object {
            "verticalAlign": "middle",
          }
        }
      >
        loading...
      </span>
    </li>
    <li>
      <button
        className="btn btn-thin btn-outline-danger"
        onClick={[Function]}
        style={
          Object {
            "border": "none",
          }
        }
      >
        <i
          className="far fa-trash-alt"
        />
      </button>
      <span
        style={
          Object {
            "verticalAlign": "middle",
          }
        }
      >
        Test address 2
      </span>
    </li>
  </ul>
</div>
`;

exports[`Render when service area targets are being shown 1`] = `
<div>
  <div
    style={
      Object {
        "backgroundColor": "#e0e0e0",
        "paddingLeft": "10px",
      }
    }
  >
    <span>
      3
       items selected
    </span>
    <button
      className="btn btn-outline-danger btn-sm float-right"
      onClick={[Function]}
      style={
        Object {
          "marginTop": "3px",
        }
      }
    >
      <i
        className="far fa-trash-alt"
      />
    </button>
  </div>
  <ul
    style={
      Object {
        "listStyleType": "none",
        "marginBottom": "0px",
        "maxHeight": "200px",
        "overflowY": "auto",
        "paddingLeft": "0px",
      }
    }
  >
    <li>
      <button
        className="btn btn-thin btn-outline-danger"
        onClick={[Function]}
        style={
          Object {
            "border": "none",
          }
        }
      >
        <i
          className="far fa-trash-alt"
        />
      </button>
      <span
        style={
          Object {
            "verticalAlign": "middle",
          }
        }
      >
        STTLWA09
      </span>
    </li>
    <li>
      <button
        className="btn btn-thin btn-outline-danger"
        onClick={[Function]}
        style={
          Object {
            "border": "none",
          }
        }
      >
        <i
          className="far fa-trash-alt"
        />
      </button>
      <span
        style={
          Object {
            "verticalAlign": "middle",
          }
        }
      >
        STTLWAXX
      </span>
    </li>
    <li>
      <button
        className="btn btn-thin btn-outline-danger"
        onClick={[Function]}
        style={
          Object {
            "border": "none",
          }
        }
      >
        <i
          className="far fa-trash-alt"
        />
      </button>
      <span
        style={
          Object {
            "verticalAlign": "middle",
          }
        }
      >
        loading...
      </span>
    </li>
  </ul>
</div>
`;
